### PR TITLE
Image Gallery deadlock fix

### PR DESCRIPTION
**Pull request by rcordovano**

### CODE REVIEW

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryOptionsPanel.java [hunk 1]
```diff
@@ -20,13 +20,11 @@
 
 import java.beans.PropertyChangeEvent;
 import java.util.EnumSet;
```
Comment (wschaeferB) [inline ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryOptionsPanel.java]: copyright date

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryTopComponent.java [hunk 2]
```diff
@@ -275,128 +270,118 @@ private void getCurrentControllerAndOpen() throws TskCoreException {
...
-                        controller.regroupDisabledProperty().addListener((Observable unused) -> Platform.runLater(() -> checkForAnalyzedGroupsForCurrentGroupBy()));
-                        controller.getGroupManager().getAnalyzedGroupsForCurrentGroupBy().addListener((Observable unused) -> Platform.runLater(() -> checkForAnalyzedGroupsForCurrentGroupBy()));
+                if (notEqual(controller, controllerForCase)) {
+                    controller = controllerForCase;
+                    /*
+                     * Create or re-create the top component's child UI
+                     * components. This is currently done every time a new
+                     * controller is created (i.e., a new case is opened).
+                     *
+                     * RJCTODO: Is it possible to just pass a new controller to
```
Comment (wschaeferB) [inline ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryTopComponent.java]: Unanswered question about alternative implementation as RJCTODO

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/actions/OpenAction.java [hunk 3]
```diff
@@ -153,10 +156,20 @@ public void performAction() {
         }
         Platform.runLater(() -> {
             ImageGalleryController controller;
-            try {
-                controller = ImageGalleryModule.getController();
```
Comment (wschaeferB) [inline ImageGallery/src/org/sleuthkit/autopsy/imagegallery/actions/OpenAction.java]: ImageGalleryModule now unused import.

Also copyright update

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/datamodel/grouping/DrawableGroup.java [hunk 4]
```diff
@@ -113,14 +120,10 @@ public DoubleBinding hashHitDensityProperty() {
      */
     public synchronized long getHashSetHitsCount() {
         if (hashSetHitsCount.get() < 0) {
-            try {
-                hashSetHitsCount.set(fileIDs.stream()
-                        .map(ImageGalleryModule.getController().getHashSetManager()::isInAnyHashSet)
```
Comment (wschaeferB) [inline ImageGallery/src/org/sleuthkit/autopsy/imagegallery/datamodel/grouping/DrawableGroup.java]: ImageGalleryModule now unused import.

also copyright update

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/gui/drawableviews/DrawableTileBase.java [hunk 5]
```diff
@@ -202,7 +202,7 @@ private ContextMenu buildContextMenu(DrawableFile file) {
                 final MenuItem extractMenuItem = new MenuItem(Bundle.DrawableTileBase_menuItem_extractFiles());
                 extractMenuItem.setOnAction(actionEvent
                         -> SwingUtilities.invokeLater(() -> {
-                            TopComponent etc = WindowManager.getDefault().findTopComponent(ImageGalleryTopComponent.PREFERRED_ID);
```
Comment (wschaeferB) [inline ImageGallery/src/org/sleuthkit/autopsy/imagegallery/gui/drawableviews/DrawableTileBase.java]: WindowManager now unused import

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/gui/drawableviews/GroupPane.java [hunk 6]
```diff
@@ -864,7 +864,7 @@ private ContextMenu buildContextMenu() {
 
             extractMenuItem.setOnAction(actionEvent -> {
                 SwingUtilities.invokeLater(() -> {
-                    TopComponent etc = WindowManager.getDefault().findTopComponent(ImageGalleryTopComponent.PREFERRED_ID);
```
Comment (wschaeferB) [inline ImageGallery/src/org/sleuthkit/autopsy/imagegallery/gui/drawableviews/GroupPane.java]: WindowManager now unused import.

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/FileIDSelectionModel.java [hunk 7]
```diff
@@ -79,7 +79,7 @@ public FileIDSelectionModel(ImageGalleryController controller) {
                 Children.Array children = new Children.Array();
                 children.add(fileNodeArray);
 
-                ImageGalleryTopComponent etc = (ImageGalleryTopComponent) WindowManager.getDefault().findTopComponent(ImageGalleryTopComponent.PREFERRED_ID);
```
[the remark this answers is not in the document]
Comment (rcordovano) [reply ImageGallery/src/org/sleuthkit/autopsy/imagegallery/FileIDSelectionModel.java]: I removed the unused import and updated the copyright.

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/datamodel/grouping/DrawableGroup.java (hunk 4 above)
Comment (rcordovano) [reply ImageGallery/src/org/sleuthkit/autopsy/imagegallery/datamodel/grouping/DrawableGroup.java]: I removed the unused import and updated the copyright.

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/gui/drawableviews/GroupPane.java (hunk 6 above)
Comment (rcordovano) [reply ImageGallery/src/org/sleuthkit/autopsy/imagegallery/gui/drawableviews/GroupPane.java]: I removed the unused import and corrected the copyright.

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/gui/drawableviews/DrawableTileBase.java (hunk 5 above)
Comment (rcordovano) [reply ImageGallery/src/org/sleuthkit/autopsy/imagegallery/gui/drawableviews/DrawableTileBase.java]: I removed the unused import and updated the copyright.

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryOptionsPanel.java (hunk 1 above)
Comment (rcordovano) [reply ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryOptionsPanel.java]: Updated / corrected copyright, added serialization ID, and removed dead logger variable.

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/actions/OpenAction.java (hunk 3 above)
Comment (rcordovano) [reply ImageGallery/src/org/sleuthkit/autopsy/imagegallery/actions/OpenAction.java]: Updated / corrected copyright and removed unused import.

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryController.java [hunk 8]
```diff
@@ -68,60 +77,264 @@
...
     /**
-     * true if Image Gallery should listen to ingest events, false if it should
-     * not listen to speed up ingest
+     * A flag that controls whether or not the image gallery controller is
+     * handling various application events. Set to true by default.
      */
-    private final SimpleBooleanProperty listeningEnabled = new SimpleBooleanProperty(false);
+    private final SimpleBooleanProperty listeningEnabled;
+
+    @ThreadConfined(type = ThreadConfined.ThreadType.JFX) // RJCTODO: Why? This does not seem to be enforced.
```
[the remark this answers is not in the document]
Comment (rcordovano) [reply ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryController.java]: Removed the RJCTODO.

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryController.java [hunk 9]
```diff
@@ -68,60 +77,264 @@
...
+
+    void startUp() throws TskCoreException {
+        selectionModel = new FileIDSelectionModel(this);
+        thumbnailCache = new ThumbnailCache(this);
 
-    private final ListeningExecutorService dbExecutor;
+        /*
+         * The next two lines need to be executed in this order.
+         *
+         * RJCTODO: Why? This suggests there is some inappropriate coupling
```
[the remark this answers is not in the document]
Comment (rcordovano) [reply ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryController.java]: Created a story and added a reference to the story to the code.

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryController.java [hunk 10]
```diff
@@ -165,94 +374,62 @@ public boolean isListeningEnabled() {
...
-
-    }
-
     /**
-     * @return Currently displayed group or null if nothing is being displayed
+     * Gets the state of the image group display area in the UI.
+     * 
+     * @return The current state. 
+     * 
+     * RJCTODO: Why do we have both getViewState and viewStateProperty methods?
```
[the remark this answers is not in the document]
Comment (rcordovano) [reply ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryController.java]: Removed RJCTODO. It's a JFX thing.

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryController.java [hunk 11]
```diff
@@ -165,94 +374,62 @@ public boolean isListeningEnabled() {
...
-     * Get observable property of the current group. The UI currently changes
-     * based on this property changing, which happens when other actions and
-     * threads call advance().
-     *
-     * @return Currently displayed group (as a property that can be observed)
+     * Gets the state of the image group display area in the UI.
+     * 
+     * @return The current state. 
+     * 
+     * RJCTODO: Why do we have both getViewState and viewStateProperty methods?
```
[the remark this answers is not in the document]
Comment (rcordovano) [reply ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryController.java]: Resolved. It's a JFX thing.

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryController.java [hunk 12]
```diff
@@ -553,16 +716,263 @@ public ReadOnlyIntegerProperty getDBTasksQueueSizeProperty() {
...
+    private class IngestModuleEventListener implements PropertyChangeListener {
+
+        @Override
+        public void propertyChange(PropertyChangeEvent event) {
+            /*
+             * For all ingest module events, call a method that enables/disables
+             * regrouping based on the drawables database task queue size and
+             * whether or not ingest is running.
+             *
+             * RJCTODO: Why do we need to call updateRegroupDisabled both if the
```
[the remark this answers is not in the document]
Comment (rcordovano) [reply ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryController.java]: Looked into this. It's actually pretty self-evident if you look at the method called by the listener. Removed RJCTODO.

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryController.java [hunk 13]
```diff
@@ -553,16 +716,263 @@ public ReadOnlyIntegerProperty getDBTasksQueueSizeProperty() {
...
+            "ImageGalleryController.dataSourceAnalyzed.confDlg.title=Image Gallery"
+        })
+        @Override
+        public void propertyChange(PropertyChangeEvent event) {
+            /*
+             * For all ingest job events, call a method that enables/disables
+             * regrouping based on the drawables database task queue size and
+             * whether or not ingest is running.
+             *
+             * RJCTODO: Why do we need to call updateRegroupDisabled both if the
```
[the remark this answers is not in the document]
Comment (rcordovano) [reply ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryController.java]: Looked into this. It's actually pretty self-evident if you look at the method called by the listener. Removed RJCTODO.

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryController.java [hunk 14]
```diff
@@ -68,60 +77,264 @@
...
 
-    public Case getAutopsyCase() {
-        return autopsyCase;
+        viewStateProperty().addListener((Observable observable) -> {
+            selectionModel.clearSelection();
+            undoManager.clear();
+        });
+
+        /*
+         * RJCTODO: Why do we need to call updateRegroupDisabled both if the
```
[the remark this answers is not in the document]
Comment (rcordovano) [reply ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryController.java]: I resolved the TODO into a comment.

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryController.java [hunk 15]
```diff
@@ -553,16 +716,263 @@ public ReadOnlyIntegerProperty getDBTasksQueueSizeProperty() {
...
+                            Content newDataSource = (Content) event.getNewValue();
+                            if (isListeningEnabled()) {
+                                drawableDB.insertOrUpdateDataSource(newDataSource.getId(), DrawableDB.DrawableDbBuildStatusEnum.UNKNOWN);
+                            }
+                        }
+                        break;
+                    case CONTENT_TAG_ADDED:
+                        final ContentTagAddedEvent tagAddedEvent = (ContentTagAddedEvent) event;
+                        long objId = tagAddedEvent.getAddedTag().getContent().getId();
+                        drawableDB.addTagCache(objId); // RJCTODO: Why add the tag to the cache before doing the in DB check?
```
[the remark this answers is not in the document]
Comment (rcordovano) [reply ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryController.java]: Created a story and added a reference to it to the code.

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryController.java [hunk 16]
```diff
@@ -553,16 +716,263 @@ public ReadOnlyIntegerProperty getDBTasksQueueSizeProperty() {
...
+                        if (drawableDB.isInDB(objId)) {
+                            tagsManager.fireTagAddedEvent(tagAddedEvent);
+                        }
+                        break;
+                    case CONTENT_TAG_DELETED:
+                        final ContentTagDeletedEvent tagDeletedEvent = (ContentTagDeletedEvent) event;
+                        if (drawableDB.isInDB(tagDeletedEvent.getDeletedTagInfo().getContentID())) {
+                            tagsManager.fireTagDeletedEvent(tagDeletedEvent);
+                        } // RJCTODO: Why not remove the tag from the cache?
```
[the remark this answers is not in the document]
Comment (rcordovano) [reply ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryController.java]: Created a story and added a reference to it to the code.

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryController.java [hunk 17]
```diff
@@ -553,16 +716,263 @@ public ReadOnlyIntegerProperty getDBTasksQueueSizeProperty() {
...
+             * or ingest module application event is published? Look at the
+             * IngestModuleEventListener and look at the the startUp method to
+             * see where we add a listener to the drawables database task queue.
+             */
+            Platform.runLater(ImageGalleryController.this::updateRegroupDisabled);
+
+            /*
+             * Only handling data source analysis events.
+             */
+            // RJCTODO: This would be less messy if IngestManager supported
```
[the remark this answers is not in the document]
Comment (rcordovano) [reply ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryController.java]: Created a story and removed the comment.

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryController.java [hunk 18]
```diff
@@ -624,6 +1034,8 @@ protected synchronized boolean isCancelled() {
 
     /**
      * task that updates one file in database with results from ingest
+     * 
+     * RJCTODO: Refactor this class into its own file.
```
[the remark this answers is not in the document]
Comment (rcordovano) [reply ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryController.java]: Refactored.

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryController.java [hunk 19]
```diff
@@ -843,6 +1257,8 @@ public void run() {
      * Grabs all files with supported image/video mime types or extensions, and
      * adds them to the Drawable DB. Uses the presence of a mimetype as an
      * approximation to 'analyzed'.
+     * 
+     * RJCTODO: Refactor this class into its own file.
```
[the remark this answers is not in the document]
Comment (rcordovano) [reply ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryController.java]: Refactored.

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/OnStart.java [hunk 20]
```diff
@@ -18,19 +18,26 @@
...
 
+import javafx.application.Platform;
+
 /**
- * The org.openide.modules.OnStart annotation tells NetBeans to invoke this
- * class's run method.
+ * An application start up task that sets a Platform property that makes the
+ * JavaFX thread continue to run until the application calls Platform.exit.
+ *
+ * RJCTODO: Is this redundant with the call to Platform.setImplicitExit(false)
```
[the remark this answers is not in the document]
Comment (rcordovano) [reply ImageGallery/src/org/sleuthkit/autopsy/imagegallery/OnStart.java]: Removed comment.

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryTopComponent.java (hunk 2 above)
Comment (rcordovano) [reply ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryTopComponent.java]: Removed comment.

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryTopComponent.java [hunk 21]
```diff
@@ -275,128 +270,118 @@ private void getCurrentControllerAndOpen() throws TskCoreException {
...
+                    controller.getGroupManager().getAnalyzedGroupsForCurrentGroupBy().addListener(groupManagerListener);
 
                     /*
-                     * Kick off a background task to query the case database for
-                     * data sources. This task may queue another task for the
-                     * JavaFX thread to allow the user to select which data
-                     * sources for which to display images. Ultimately, a task
-                     * will be queued for the AWT EDT that will show the top
-                     * component window.
+                     * RJCTODO: Should this be done in
```
[the remark this answers is not in the document]
Comment (rcordovano) [reply ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryTopComponent.java]: Removed comment.

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryTopComponent.java [hunk 22]
```diff
@@ -275,128 +270,118 @@ private void getCurrentControllerAndOpen() throws TskCoreException {
...
+                         * If there is only one datasource or the grouping
+                         * criterion is already set to something other than by
+                         * path (the default), proceed to open this top
+                         * component. Otherwise, do a dialog to allow the user
+                         * to select the data sources for which images are to be
+                         * displayed, then open the top component.
+                         */
+                        List<DataSource> dataSources = controller.getCaseDatabase().getDataSources();
+                        Map<DataSource, Boolean> dataSourcesWithTooManyFiles = new HashMap<>();
+                        // RJCTODO: At least some of this designation of "all data sources" with null seems uneccessary;
```
[the remark this answers is not in the document]
Comment (rcordovano) [reply ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryTopComponent.java]: Created a story and added reference to code.

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryTopComponent.java [hunk 23]
```diff
@@ -454,6 +439,8 @@ public Lookup getLookup() {
      * manager and removes the blocking progress spinner if there are analyzed
      * groups; otherwise adds a blocking progress spinner with an appropriate
      * message.
+     *
+     * RJCTODO: Is this an accurate method description?
```
[the remark this answers is not in the document]
Comment (rcordovano) [reply ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryTopComponent.java]: Removed comment.

File: ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryTopComponent.java [hunk 24]
```diff
@@ -468,59 +455,58 @@ public Lookup getLookup() {
...
-            // @@@ Need to check timing on this and make sure we have only groups for the selected DS.  Seems like rebuild can cause groups to be created for a DS that is not later selected...
-            // RJCTODO: Get Brian's TODO resolved.
-            if (isNotEmpty(groupManager.getAnalyzedGroupsForCurrentGroupBy())) {
-                clearNotification();
-                return;
+        GroupManager groupManager = controller.getGroupManager();
+
+        // if there are groups to display, then display them
+        // @@@ Need to check timing on this and make sure we have only groups for the selected DS.  Seems like rebuild can cause groups to be created for a DS that is not later selected...
+        // RJCTODO: Get Brian's TODO resolved.
```
[the remark this answers is not in the document]
Comment (rcordovano) [reply ImageGallery/src/org/sleuthkit/autopsy/imagegallery/ImageGalleryTopComponent.java]: Created a story and referenced it in the code.

**Comment by rcordovano** [comment]
@wschaeferB, I addressed the Codacy issues (we'll see what the next analysis says) and your comments.

**Pull request merged**